core/aggsigdb/memory_v2.go

```diff
@@ -0,0 +1,144 @@
+// Copyright © 2022-2024 Obol Labs Inc. Licensed under the terms of a Business Source License 1.1
+
+package aggsigdb
+
+import (
+	"context"
+	"fmt"
+	"runtime"
+	"sync"
+
+	"github.com/obolnetwork/charon/app/errors"
+	"github.com/obolnetwork/charon/app/log"
+	"github.com/obolnetwork/charon/app/z"
+	"github.com/obolnetwork/charon/core"
+)
+
+// MemDBV2 is a basic memory implementation of core.AggSigDB.
+type MemDBV2 struct {
+	data       sync.Map // map[memDBKey]core.SignedData
+	keysByDuty sync.Map // map[core.Duty][]memDBKey,  Key index by duty for fast deletion.
+	deadliner  core.Deadliner
+	closed     chan struct{}
+}
+
+// NewMemDBV2 creates a basic memory based AggSigDB.
+func NewMemDBV2(deadliner core.Deadliner) *MemDBV2 {
+	return &MemDBV2{
+		// data, keysByDuty are okay to use without explicit initialization
+		deadliner: deadliner,
+		closed:    make(chan struct{}),
+	}
+}
+
+func (m *MemDBV2) store(duty core.Duty, pubKey core.PubKey, data core.SignedData) error {
+	data, err := data.Clone()
+	if err != nil {
+		return err
+	}
+
+	_ = m.deadliner.Add(duty) // TODO(corver): Distinguish between no deadline supported vs already expired.
+
+	key := memDBKey{duty, pubKey}
+
+	if rawExisting, ok := m.data.Load(key); ok {
+		existing, ok := rawExisting.(core.SignedData)
+		if !ok {
+			return errors.New("data stored in aggsigdb not of core.SignedData type", z.Str("key", fmt.Sprintf("%+v", key)))
+		}
+
+		equal, err := dataEqual(existing, data)
+		if err != nil {
+			return err
+		} else if !equal {
+			return errors.New("mismatching data")
+		}
+	} else {
+		m.data.Store(key, data)
+		rawKbd, _ := m.keysByDuty.Load(duty)
+
+		if rawKbd == nil {
+			rawKbd = []memDBKey{}
+		}
+
+		kbd, ok := rawKbd.([]memDBKey)
+		if !ok {
+			return errors.New("indexing key data stored in aggsigdb not of []memDBKey type", z.Str("duty", duty.String()))
+		}
+
+		kbd = append(kbd, key)
+		m.keysByDuty.Store(duty, kbd)
+	}
+
+	return nil
+}
+
+func (m *MemDBV2) Store(ctx context.Context, duty core.Duty, set core.SignedDataSet) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-m.closed:
+		return ErrStopped
+	default:
+	}
+
+	for pubKey, data := range set {
+		if err := m.store(duty, pubKey, data); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (m *MemDBV2) Await(ctx context.Context, duty core.Duty, pubKey core.PubKey) (core.SignedData, error) {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case <-m.closed:
+			return nil, ErrStopped
+		default:
+			maybeDataRaw, ok := m.data.Load(memDBKey{duty, pubKey})
```

Comment:
> This busy waiting is a big concern.

For what reason?

Considering this function call is constrained by the context, is it really a concern?

The context establishes cancellation, which in our codebase is always tied to a `context.WithTimeout`.

Reply:
> this does not look like a natural pattern to golang IMO

I see your point. 

The behavior you're describing is pretty similar to the original implementation, which while is impressively Go-esque is also kind of confusing to look at.

Considering `AggSigDB` is just a fancy map with an `Await` function, I'm keen on having the simplest implementation possible which exhibits the same behavior (as far as our testing goes).

Reply:
I think your implementation is fine, except for the missing necessary locks for maps/data. And, maybe instead of `Gosched` we could be introducing an interval and Sleep that interval between attempts. Or, have a single dedicated channel that will signal "any" data change, which is more elegant.

Reply:
I'm keen on leaving `Gosched` there to allow other goroutines to be scheduled, and also leaving `Await` to return as soon as possible, since it's core to the lifecycle.

Maybe let's test this and see if it makes a difference CPU-wise? Marking ourselves as schedule-able should be enough to avoid trashing and IMO it's better than sleeping.

Reply:
Imagine if you have a single shared notification channel, in this case you don't need to block Await, no need to Gosched, etc. Just send struct{}{} whenever you Store a value. All Awaits will get it immediately and go sleep again if there is no value of interest. Isn't this cleaner?

Reply:
Let me play with this a little idea then

Reply:
I think we need this: https://pkg.go.dev/sync#Cond as a signaling mechanism instead of sleeping, busy-waiting, polling.

Reply:
Condvars are tricky, as the doc says, it might be better to sync with channels.

I pushed a candidate implementation that implements what you suggested, let me know what you think.